agent/engine/engine_sudo_linux_integ_test.go

```diff
@@ -191,206 +164,84 @@ func createTestLocalVolumeMountTask() *apitask.Task {
-	server.Start()
-	defer server.Close()
+	cfg.Cluster = TestCluster
 
-	port := getPortFromAddr(t, server.URL)
-	ec2MetadataClient, err := ec2.NewEC2MetadataClient(nil)
-	require.NoError(t, err)
-	serverURL := fmt.Sprintf("http://%s:%s", getHostPrivateIP(t, ec2MetadataClient), port)
-	defer setV3MetadataURLFormat(serverURL + "/v3/%s")()
-
-	testTask := createFirelensTask(t)
+	// Setup task engine
+	taskEngine, done, _, _ := SetupIntegTestTaskEngine(cfg, nil, t)
 	taskEngine.(*DockerTaskEngine).resourceFields = &taskresource.ResourceFields{
 		ResourceFieldsCommon: &taskresource.ResourceFieldsCommon{
-			EC2InstanceID: testInstanceID,
+			EC2InstanceID: TestInstanceID,
 		},
 	}
-	go taskEngine.AddTask(testTask)
-	testEvents := InitTestEventCollection(taskEngine)
-
-	//Verify logsender container is running
-	VerifyContainerStatus(apicontainerstatus.ContainerRunning, testTask.Arn+":logsender", testEvents, t)
-
-	//Verify firelens container is running
-	VerifyContainerStatus(apicontainerstatus.ContainerRunning, testTask.Arn+":firelens", testEvents, t)
-
-	//Verify task is in running state
-	VerifyTaskStatus(apitaskstatus.TaskRunning, testTask.Arn, testEvents, t)
-
-	//Verify logsender container is stopped
-	VerifyContainerStatus(apicontainerstatus.ContainerStopped, testTask.Arn+":logsender", testEvents, t)
-
-	//Verify firelens container is stopped
-	VerifyContainerStatus(apicontainerstatus.ContainerStopped, testTask.Arn+":firelens", testEvents, t)
-
-	//Verify the task itself has stopped
-	VerifyTaskStatus(apitaskstatus.TaskStopped, testTask.Arn, testEvents, t)
+	defer done()
 
-	taskID := testTask.GetID()
+	// Setup task metadata server
+	setupTaskMetadataServer(t)
 
-	//declare a cloudwatch client
-	awsCfg, err := awsconfig.LoadDefaultConfig(context.TODO(), awsconfig.WithRegion(testECSRegion))
-	require.NoError(t, err, "Unable to load AWS config")
-	cwlClient := cloudwatchlogs.NewFromConfig(awsCfg)
-	params := &cloudwatchlogs.GetLogEventsInput{
-		LogGroupName:  aws.String(testLogGroupName),
-		LogStreamName: aws.String(fmt.Sprintf("firelens-fluentbit-logsender-firelens-%s", taskID)),
+	testCases := []struct {
+		name                  string
+		firelensContainerUser string
+	}{
+		{
+			name: "default",
+		},
+		{
+			name:                  "root-user",
+			firelensContainerUser: "0:1000",
+		},
+		{
+			name:                  "non-root-user",
+			firelensContainerUser: "1234:5678",
+		},
 	}
 
-	// wait for the cloud watch logs
-	resp, err := waitCloudwatchLogs(cwlClient, params)
-	require.NoError(t, err)
-	// there should only be one event as we are echoing only one thing that part of the include-filter
-	assert.Equal(t, 1, len(resp.Events))
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// Create a temporary directory to send app logs to
+			tmpDir := createFirelensLogDir(t, tc.name, tc.firelensContainerUser)
+			defer func() {
+				err := os.RemoveAll(tmpDir)
+				require.NoError(t, err)
+			}()
```

Comment:
Is there a reason not to move this into the function usage to avoid someone calling the method without cleanup?

Reply:
Good q - defer statements in go execute after the method in which they are return back to the caller. In this case, moving the defer inside a helper method would execute it as soon as the helper returns. I'd like to remove the temp directory only once each test case is executed completely.

Reply:
https://go.dev/tour/flowcontrol/12